### PR TITLE
Bug fixes / QoL improvements

### PR DESCRIPTION
**Pull request by libraun**
Fixes:

+ Fix to issue #235: VT242 Terminal now accepts "jtitor" as a (wrong) password.
+ Fix to issue #226: The game now recognizes "remove hinges", and uses the same callback as "remove pins"
+ Fix to issue #224: Added new attribute to the game, "hint_given", and added it to every location set to False by default.
+ Fix to issue #213: Added cheap scenery to represent April's office door to hallway 5
+ Fix to issue #230: Added a check to determine whether or not player has flathead screwdriver prior to "revealing" it again.
+ Progress on issue #238: Added more door cheap scenery, and added a clause to indicate that one of two doors in a certain hallway is already open (to avoid confusion upon typing "look door")

+ Typo fixes

Other Additions:

+ Added fire escape to lobby east
+ Added cheap scenery to hallway3

### CODE REVIEW

**Comment by pahp** [comment]
Hi @libraun --

Thanks for this work! I did some testing, and your notes gave me an idea...

```
> remove pins
No need to concern yourself with that.
```

I'm not sure that this is better than `That wouldn't be useful.` ...?

Seems like maybe in addition to `remove (hinges|pins) with flathead` it should say something if you type `remove (hinges|pins)` -- like `Remove them with what?` -- is that straightforward to do, @spacehobo ?

**Comment by pahp** [comment]
Another thing @libraun -- I think the emergency exit door should be present (and thanks for making it!), but when you try to go through it, it tells you to wait until you've completed things. As implemented, it lets you into a new room (the top of the stairs?), but that room goes nowhere... it breaks disbelief; why can't I just try to go downstairs... OMG where are the stairs?

So again, I think that room should be inaccessible until the tasks are complete, and then once you are let through there is a fire escape that works however we intend for it to work. Thanks!

**Comment by pahp** [comment]
@libraun -- [See comment on PR](https://github.com/UMDLARS/one_night_in_sf/pull/227#issuecomment-1981393747). I don't think you should be able to go into the fire escape room until you have all the items. It is weird to go into a room that has no visible exits. Also, presumably the fire exit would have stairs, but they're not visible (because we don't want them to use them). 

It would be cleaner to simply stop them from going through the fire exit door unless they have all the items.

I also think that the game should not be in a broken state, so I think that once the player has all the items, if they try to use the door, it just directs them to the elevator for now ("Do you really want to carry an IBM 5100 down 13 flights of stairs?"). We can add in some kind of funny fire escape thing later if we want.

So...

 1. If they try to use the fire escape door, it should say no if they do not have all the items.
 2. If they do have all the items, it should tell them to use the elevator in a funny way.

Thanks!

**Comment by libraun** [comment]
Thanks for these notes!! Per the fire escape: I actually caught that right
before you emailed me this; I thought it was weird that the player opens
the door  (to “inspect” the fire escape) without closing it again, so I
made an edit which clarifies that. Additionally there is a check in there
that switches up the dialogue, if the player has MAX_SCORE.

I’ll go through these more at home, and get up to date. Thanks again!

**Comment by libraun** [comment]
I was also just looking at how extension works in PunyInform, so I’m still
curious about that last one.

**Comment by spacehobo** [comment]
This is kind of a huge pull request that lumps a lot of things together, so it's going to be difficult to review.  It's really best practice to keep commits and PRs small, and maybe make one per issue where possible.

The first thing I'm not sure about is the binding of `remove` to `ThrowAt`, but in a pattern that doesn't have any indirect objects.  At best this needs comments, but at worst it seems like it may be the result of a misunderstanding.  Can you help me understand why that's there?

Second, I don't see any tests for any of the added functionality.  We need to be able to just run `make` to verify that it does what you think it does!  The fire escape absolutely needs tests before we can merge this.

Third, the hinge puzzle is either getting complex enough that it needs an actual door object, or it needs more logic to display things differently once the pins are removed.

**Comment by libraun** [comment]
Oops, that “remove” binding shouldn’t have made it in there. Sorry for all
the confusion, I was worried I might’ve made this one way too big. I’ll go
through and try to make a log of all the changes for this PR. I’ll try to
make it as complete as possible, but a lot of the changes were just typo
fixes, and some rewordings to make things more apparent where needed.

I don’t suppose there’s a way to “split” pull requests into commits?

**Comment by libraun** [comment]
Essentially, the “remove” binding was just to test and see if items could
be removed from the player’s inventory. It didn’t work out, so disregard
that portion.

I’ll also add a good amount of comments, hopefully that should make it
easier to disambiguate.

**Comment by spacehobo** [comment]
https://github.com/UMDLARS/one_night_in_sf/pull/227/files#diff-a027df5b87ceff2dad568d6fcdb3ef3a22d89cd212a050a936330731221c6732R850 ← For this test, I would test for `has moved` rather than whether it's currently in the player's inventory.

File: untitledHeistGame.inf [hunk 1]
```diff
@@ -773,17 +838,24 @@ u_to [;
...
 with
 	name 'tarp' 'tarpaulin' 'PET' 'polyethylene',
-	description "This is a blue polyethylene tarp, and its craftsmanship is clearly of the highest quality. On it are spatterings of off-white paint.",
+	description "This is a blue polyethylene tarp, the craftsmanship of which is clearly of the highest quality. On it are spatterings of off-white paint.",
 	before[;
 		Take, Transfer:
-			print "As you fold up and take the tarp you notice underneath it there was a flathead screwdriver!";
-			give flatHead ~concealed;
-			return 1;
+			if (flatHead in player) {
```
Comment (spacehobo) [inline untitledHeistGame.inf:850]: I would test `if (flathead has concealed) {` instead, here.

File: untitledHeistGame.inf [hunk 2]
```diff
@@ -883,6 +968,7 @@ e_to Hallway3,
 s_to Kitchenette,
 n_to Office3Door,
 
+has hint_given ~hint_given,
```
Comment (spacehobo) [inline untitledHeistGame.inf:971]: When you do this, you're giving the object *both* `hint_given` and *not* `hint_given`.  Attributes default to false, so you shouldn't need to pre-populate these.

File: untitledHeistGame.inf [hunk 3]
```diff
@@ -473,10 +528,10 @@ with
 					if (self has general && noun == 'tipler') {
 						give self ~locked;
 						Achieved(GET_TITOR_PLANS);
-						print "The system pauses ominously for a moment, and then presents you with a command prompt.  Poking around briefly, you discover a directory full of technical schematics and research papers.^You start printing the data which begins exiting out the machine on a spool of green bar paper. You grab the printout.";
+						print "The system pauses ominously for a moment, and then presents you with a command prompt.  Poking around briefly, you discover a directory full of technical schematics and research papers.^You start printing the data, which begins exiting out of the machine onto a spool of green bar paper. You grab the printout.";
 						move printout to player;
 						return true;
-					} else if (noun == 'jtitor') {
+					} else if (self hasnt general && noun == 'jtitor') {
```
Comment (spacehobo) [inline untitledHeistGame.inf:534]: Oh, very good catch!

**Comment by spacehobo** [comment]
>  I don’t suppose there’s a way to “split” pull requests into commits?

Mainly the way to do it I think is to make new branches and cherry-pick commits into those branches, and propose each one separately.  I tend to name my branches after issue numbers, and begin by adding a failing test for that issue.

**Comment by libraun** [comment]
That’s smart, I’ll see what I can do there at this point in terms of moving
commits around (and maybe splicing them into different PR’s if possible).
Definitely noted for the future.

File: untitledHeistGame.inf (hunk 1 above)
Comment (libraun) [reply untitledHeistGame.inf]: Done!

File: untitledHeistGame.inf (hunk 2 above)
Comment (libraun) [reply untitledHeistGame.inf:969]: Thank you, I was wondering about this.

File: untitledHeistGame.inf (hunk 3 above)
Comment (libraun) [reply untitledHeistGame.inf:531]: Courtesy of an Advanced Computer Security student!

**Comment by libraun** [comment]
> Another thing @libraun -- I think the emergency exit door should be present (and thanks for making it!), but when you try to go through it, it tells you to wait until you've completed things. As implemented, it lets you into a new room (the top of the stairs?), but that room goes nowhere... it breaks disbelief; why can't I just try to go downstairs... OMG where are the stairs?
> 
> So again, I think that room should be inaccessible until the tasks are complete, and then once you are let through there is a fire escape that works however we intend for it to work. Thanks!

Working on this now, should I go ahead and pull these changes into this branch or wait until after the commit? It should only take a couple adjustments to the test file.

**Comment by libraun** [comment]
> Hi @libraun --
> 
> Thanks for this work! I did some testing, and your notes gave me an idea...
> 
> ```
> > remove pins
> No need to concern yourself with that.
> ```
> 
> I'm not sure that this is better than `That wouldn't be useful.` ...?
> 
> Seems like maybe in addition to `remove (hinges|pins) with flathead` it should say something if you type `remove (hinges|pins)` -- like `Remove them with what?` -- is that straightforward to do, @spacehobo ?

Definitely agreed, and fixed! Not sure how that happened. Not sure about the "remove pins" with no second phrase -- I tried checking for an empty string as second with no success.

**Comment by spacehobo** [comment]
It's looking much better!  Remember to test "failures" as well, so that the user tries things that shouldn't work, and you prove that they continue to fail even as we change code.

Also I mentioned on Signal, but you can see the way the `include` directive can work to let you create a complex game state to use at the start of multiple tests.

**Comment by libraun** [comment]
> It's looking much better! Remember to test "failures" as well, so that the user tries things that shouldn't work, and you prove that they continue to fail even as we change code.
> 
> Also I mentioned on Signal, but you can see the way the `include` directive can work to let you create a complex game state to use at the start of multiple tests.

Wonderful! And testing failures is a great policy that I need to remember to use.

**Comment by libraun** [comment]
> Another thing @libraun -- I think the emergency exit door should be present (and thanks for making it!), but when you try to go through it, it tells you to wait until you've completed things. As implemented, it lets you into a new room (the top of the stairs?), but that room goes nowhere... it breaks disbelief; why can't I just try to go downstairs... OMG where are the stairs?
> 
> So again, I think that room should be inaccessible until the tasks are complete, and then once you are let through there is a fire escape that works however we intend for it to work. Thanks!

@pahp  It might've been mentioned in Slack, but how is this looking as cheap scenery instead of as it was (a separate room)? Regardless, I understand how you mean by funkiness in terms of opening the door to look into the fire escape. It's disorienting.

**Pull request merged**